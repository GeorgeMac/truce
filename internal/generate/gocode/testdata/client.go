// DO NOT EDIT.
// This code was generated by Truce.
package example

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"io/ioutil"
	"net/http"
	"net/url"
)

var _ = bytes.Compare

type ExampleClient struct {
	client *http.Client
	host   *url.URL
}

func NewExampleClient(host string) (*ExampleClient, error) {
	u, err := url.Parse(host)
	if err != nil {
		return nil, err
	}

	return &ExampleClient{client: http.DefaultClient, host: u}, nil
}

func (_c *ExampleClient) GetPost(ctxt context.Context, id string) (Post, error) {
	_u, _err := _c.host.Parse(fmt.Sprintf("/api/v1/posts/%v", id))
	if _err != nil {
		return
	}

	var (
		_body io.Reader
		_resp *http.Response
	)

	_req, _err := http.NewRequest("GET", _u.String(), _body)
	if _err != nil {

		return Post{}, _err

	}

	_resp, _err = c.client.Do(_req.WithContext(ctxt))
	if _err != nil {

		return Post{}, _err

	}

	defer func() {
		_, _ = io.Copy(ioutil.Discard, _resp.Body)
		_ = _resp.Body.Close()
	}()

	if _err = _checkResponse(_resp); _err != nil {

		return Post{}, _err

	}

	_err = json.NewDecoder(resp.Body).Decode(&rtn)
	return _rtn, _err

}
func (_c *ExampleClient) GetPosts(ctxt context.Context) ([]Post, error) {
	_u, _err := _c.host.Parse("/api/v1/posts")
	if _err != nil {
		return
	}

	var (
		_body io.Reader
		_resp *http.Response
	)

	_req, _err := http.NewRequest("GET", _u.String(), _body)
	if _err != nil {

		return []Post{}, _err

	}

	_resp, _err = c.client.Do(_req.WithContext(ctxt))
	if _err != nil {

		return []Post{}, _err

	}

	defer func() {
		_, _ = io.Copy(ioutil.Discard, _resp.Body)
		_ = _resp.Body.Close()
	}()

	if _err = _checkResponse(_resp); _err != nil {

		return []Post{}, _err

	}

	_err = json.NewDecoder(resp.Body).Decode(&rtn)
	return _rtn, _err

}
func (_c *ExampleClient) GetUser(ctxt context.Context, id string) (User, error) {
	_u, _err := _c.host.Parse(fmt.Sprintf("/api/v1/users/%v", id))
	if _err != nil {
		return
	}

	var (
		_body io.Reader
		_resp *http.Response
	)

	_req, _err := http.NewRequest("GET", _u.String(), _body)
	if _err != nil {

		return User{}, _err

	}

	_resp, _err = c.client.Do(_req.WithContext(ctxt))
	if _err != nil {

		return User{}, _err

	}

	defer func() {
		_, _ = io.Copy(ioutil.Discard, _resp.Body)
		_ = _resp.Body.Close()
	}()

	if _err = _checkResponse(_resp); _err != nil {

		return User{}, _err

	}

	_err = json.NewDecoder(resp.Body).Decode(&rtn)
	return _rtn, _err

}
func (_c *ExampleClient) GetUsers(ctxt context.Context) ([]User, error) {
	_u, _err := _c.host.Parse("/api/v1/users")
	if _err != nil {
		return
	}

	var (
		_body io.Reader
		_resp *http.Response
	)

	_req, _err := http.NewRequest("GET", _u.String(), _body)
	if _err != nil {

		return []User{}, _err

	}

	_resp, _err = c.client.Do(_req.WithContext(ctxt))
	if _err != nil {

		return []User{}, _err

	}

	defer func() {
		_, _ = io.Copy(ioutil.Discard, _resp.Body)
		_ = _resp.Body.Close()
	}()

	if _err = _checkResponse(_resp); _err != nil {

		return []User{}, _err

	}

	_err = json.NewDecoder(resp.Body).Decode(&rtn)
	return _rtn, _err

}
func (_c *ExampleClient) PatchPost(ctxt context.Context, id string, post PatchPostRequest) (Post, error) {
	_u, _err := _c.host.Parse(fmt.Sprintf("/api/v1/posts/%v", id))
	if _err != nil {
		return
	}

	var (
		_body io.Reader
		_resp *http.Response
	)

	_buf := &bytes.Buffer{}
	_body = _buf
	if err = json.NewEncoder(_buf).Encode(post); err != nil {
		return
	}

	_req, _err := http.NewRequest("PATCH", _u.String(), _body)
	if _err != nil {

		return Post{}, _err

	}

	_resp, _err = c.client.Do(_req.WithContext(ctxt))
	if _err != nil {

		return Post{}, _err

	}

	defer func() {
		_, _ = io.Copy(ioutil.Discard, _resp.Body)
		_ = _resp.Body.Close()
	}()

	if _err = _checkResponse(_resp); _err != nil {

		return Post{}, _err

	}

	_err = json.NewDecoder(resp.Body).Decode(&rtn)
	return _rtn, _err

}
func (_c *ExampleClient) PatchUser(ctxt context.Context, id string, user PatchUserRequest) (User, error) {
	_u, _err := _c.host.Parse(fmt.Sprintf("/api/v1/users/%v", id))
	if _err != nil {
		return
	}

	var (
		_body io.Reader
		_resp *http.Response
	)

	_buf := &bytes.Buffer{}
	_body = _buf
	if err = json.NewEncoder(_buf).Encode(user); err != nil {
		return
	}

	_req, _err := http.NewRequest("PATCH", _u.String(), _body)
	if _err != nil {

		return User{}, _err

	}

	_resp, _err = c.client.Do(_req.WithContext(ctxt))
	if _err != nil {

		return User{}, _err

	}

	defer func() {
		_, _ = io.Copy(ioutil.Discard, _resp.Body)
		_ = _resp.Body.Close()
	}()

	if _err = _checkResponse(_resp); _err != nil {

		return User{}, _err

	}

	_err = json.NewDecoder(resp.Body).Decode(&rtn)
	return _rtn, _err

}
func (_c *ExampleClient) PutPost(ctxt context.Context, post PutPostRequest) (Post, error) {
	_u, _err := _c.host.Parse("/api/v1/posts")
	if _err != nil {
		return
	}

	var (
		_body io.Reader
		_resp *http.Response
	)

	_buf := &bytes.Buffer{}
	_body = _buf
	if err = json.NewEncoder(_buf).Encode(post); err != nil {
		return
	}

	_req, _err := http.NewRequest("PUT", _u.String(), _body)
	if _err != nil {

		return Post{}, _err

	}

	_resp, _err = c.client.Do(_req.WithContext(ctxt))
	if _err != nil {

		return Post{}, _err

	}

	defer func() {
		_, _ = io.Copy(ioutil.Discard, _resp.Body)
		_ = _resp.Body.Close()
	}()

	if _err = _checkResponse(_resp); _err != nil {

		return Post{}, _err

	}

	_err = json.NewDecoder(resp.Body).Decode(&rtn)
	return _rtn, _err

}
func (_c *ExampleClient) PutUser(ctxt context.Context, user PutUserRequest) (User, error) {
	_u, _err := _c.host.Parse("/api/v1/users")
	if _err != nil {
		return
	}

	var (
		_body io.Reader
		_resp *http.Response
	)

	_buf := &bytes.Buffer{}
	_body = _buf
	if err = json.NewEncoder(_buf).Encode(user); err != nil {
		return
	}

	_req, _err := http.NewRequest("PUT", _u.String(), _body)
	if _err != nil {

		return User{}, _err

	}

	_resp, _err = c.client.Do(_req.WithContext(ctxt))
	if _err != nil {

		return User{}, _err

	}

	defer func() {
		_, _ = io.Copy(ioutil.Discard, _resp.Body)
		_ = _resp.Body.Close()
	}()

	if _err = _checkResponse(_resp); _err != nil {

		return User{}, _err

	}

	_err = json.NewDecoder(resp.Body).Decode(&rtn)
	return _rtn, _err

}

func _checkResponse(resp *http.Response) error {
	switch resp.StatusCode {
	case http.StatusOK:
		return nil
	case 401:
		v := NotAuthorized{}
		if err := json.NewDecoder(resp.Body).Decode(&v); err != nil {
			return err
		}

		return v
	case 404:
		v := NotFound{}
		if err := json.NewDecoder(resp.Body).Decode(&v); err != nil {
			return err
		}

		return v
	default:
		return errors.New("unexpected status code")
	}
}
