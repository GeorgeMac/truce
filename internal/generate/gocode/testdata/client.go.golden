package example

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"io/ioutil"
	"net/http"
	"net/url"
	"time"
)

var _ = bytes.Compare
var _ = time.After

type ExampleClient struct {
	client *http.Client
	host   *url.URL
}

func NewExampleClient(host string) (*ExampleClient, error) {
	u, err := url.Parse(host)
	if err != nil {
		return nil, err
	}

	return &ExampleClient{client: http.DefaultClient, host: u}, nil
}

func (c *ExampleClient) GetPost(ctxt context.Context, v0 string) (rtn Post, err error) {
	u, err := c.host.Parse(fmt.Sprintf("/api/v1/posts/%v", v0))
	if err != nil {
		return
	}

	var (
		body io.Reader
		resp *http.Response
	)

	req, err := http.NewRequest("GET", u.String(), body)
	if err != nil {
		return
	}

	resp, err = c.client.Do(req.WithContext(ctxt))
	if err != nil {
		return
	}

	defer func() {
		_, _ = io.Copy(ioutil.Discard, resp.Body)
		_ = resp.Body.Close()
	}()

	if err = checkResponse(resp); err != nil {
		return
	}

	err = json.NewDecoder(resp.Body).Decode(&rtn)

	return
}
func (c *ExampleClient) GetPosts(ctxt context.Context) (rtn []Post, err error) {
	u, err := c.host.Parse("/api/v1/posts")
	if err != nil {
		return
	}

	var (
		body io.Reader
		resp *http.Response
	)

	req, err := http.NewRequest("GET", u.String(), body)
	if err != nil {
		return
	}

	resp, err = c.client.Do(req.WithContext(ctxt))
	if err != nil {
		return
	}

	defer func() {
		_, _ = io.Copy(ioutil.Discard, resp.Body)
		_ = resp.Body.Close()
	}()

	if err = checkResponse(resp); err != nil {
		return
	}

	err = json.NewDecoder(resp.Body).Decode(&rtn)

	return
}
func (c *ExampleClient) GetUser(ctxt context.Context, v0 string) (rtn User, err error) {
	u, err := c.host.Parse(fmt.Sprintf("/api/v1/users/%v", v0))
	if err != nil {
		return
	}

	var (
		body io.Reader
		resp *http.Response
	)

	req, err := http.NewRequest("GET", u.String(), body)
	if err != nil {
		return
	}

	resp, err = c.client.Do(req.WithContext(ctxt))
	if err != nil {
		return
	}

	defer func() {
		_, _ = io.Copy(ioutil.Discard, resp.Body)
		_ = resp.Body.Close()
	}()

	if err = checkResponse(resp); err != nil {
		return
	}

	err = json.NewDecoder(resp.Body).Decode(&rtn)

	return
}
func (c *ExampleClient) GetUsers(ctxt context.Context) (rtn []User, err error) {
	u, err := c.host.Parse("/api/v1/users")
	if err != nil {
		return
	}

	var (
		body io.Reader
		resp *http.Response
	)

	req, err := http.NewRequest("GET", u.String(), body)
	if err != nil {
		return
	}

	resp, err = c.client.Do(req.WithContext(ctxt))
	if err != nil {
		return
	}

	defer func() {
		_, _ = io.Copy(ioutil.Discard, resp.Body)
		_ = resp.Body.Close()
	}()

	if err = checkResponse(resp); err != nil {
		return
	}

	err = json.NewDecoder(resp.Body).Decode(&rtn)

	return
}
func (c *ExampleClient) PatchPost(ctxt context.Context, v0 string, v1 PatchPostRequest) (rtn Post, err error) {
	u, err := c.host.Parse(fmt.Sprintf("/api/v1/posts/%v", v0))
	if err != nil {
		return
	}

	var (
		body io.Reader
		resp *http.Response
	)

	buf := &bytes.Buffer{}
	body = buf
	if err = json.NewEncoder(buf).Encode(v1); err != nil {
		return
	}

	req, err := http.NewRequest("PATCH", u.String(), body)
	if err != nil {
		return
	}

	resp, err = c.client.Do(req.WithContext(ctxt))
	if err != nil {
		return
	}

	defer func() {
		_, _ = io.Copy(ioutil.Discard, resp.Body)
		_ = resp.Body.Close()
	}()

	if err = checkResponse(resp); err != nil {
		return
	}

	err = json.NewDecoder(resp.Body).Decode(&rtn)

	return
}
func (c *ExampleClient) PatchUser(ctxt context.Context, v0 string, v1 PatchUserRequest) (rtn User, err error) {
	u, err := c.host.Parse(fmt.Sprintf("/api/v1/users/%v", v0))
	if err != nil {
		return
	}

	var (
		body io.Reader
		resp *http.Response
	)

	buf := &bytes.Buffer{}
	body = buf
	if err = json.NewEncoder(buf).Encode(v1); err != nil {
		return
	}

	req, err := http.NewRequest("PATCH", u.String(), body)
	if err != nil {
		return
	}

	resp, err = c.client.Do(req.WithContext(ctxt))
	if err != nil {
		return
	}

	defer func() {
		_, _ = io.Copy(ioutil.Discard, resp.Body)
		_ = resp.Body.Close()
	}()

	if err = checkResponse(resp); err != nil {
		return
	}

	err = json.NewDecoder(resp.Body).Decode(&rtn)

	return
}
func (c *ExampleClient) PutPost(ctxt context.Context, v0 PutPostRequest) (rtn Post, err error) {
	u, err := c.host.Parse("/api/v1/posts")
	if err != nil {
		return
	}

	var (
		body io.Reader
		resp *http.Response
	)

	buf := &bytes.Buffer{}
	body = buf
	if err = json.NewEncoder(buf).Encode(v0); err != nil {
		return
	}

	req, err := http.NewRequest("PUT", u.String(), body)
	if err != nil {
		return
	}

	resp, err = c.client.Do(req.WithContext(ctxt))
	if err != nil {
		return
	}

	defer func() {
		_, _ = io.Copy(ioutil.Discard, resp.Body)
		_ = resp.Body.Close()
	}()

	if err = checkResponse(resp); err != nil {
		return
	}

	err = json.NewDecoder(resp.Body).Decode(&rtn)

	return
}
func (c *ExampleClient) PutUser(ctxt context.Context, v0 PutUserRequest) (rtn User, err error) {
	u, err := c.host.Parse("/api/v1/users")
	if err != nil {
		return
	}

	var (
		body io.Reader
		resp *http.Response
	)

	buf := &bytes.Buffer{}
	body = buf
	if err = json.NewEncoder(buf).Encode(v0); err != nil {
		return
	}

	req, err := http.NewRequest("PUT", u.String(), body)
	if err != nil {
		return
	}

	resp, err = c.client.Do(req.WithContext(ctxt))
	if err != nil {
		return
	}

	defer func() {
		_, _ = io.Copy(ioutil.Discard, resp.Body)
		_ = resp.Body.Close()
	}()

	if err = checkResponse(resp); err != nil {
		return
	}

	err = json.NewDecoder(resp.Body).Decode(&rtn)

	return
}

func checkResponse(resp *http.Response) error {
	switch resp.StatusCode {
	case http.StatusOK:
		return nil
	case 401:
		v := NotAuthorized{}
		if err := json.NewDecoder(resp.Body).Decode(&v); err != nil {
			return err
		}

		return v
	case 404:
		v := NotFound{}
		if err := json.NewDecoder(resp.Body).Decode(&v); err != nil {
			return err
		}

		return v

	default:
		return errors.New("unexpected status code")
	}
}
